core/src/main/scala/io/projectglow/vcf/BigVCFDatasource.scala

```diff
@@ -47,17 +47,26 @@ object BigVCFDatasource extends HlsUsageLogging {
     val rdd = data.queryExecution.toRdd
 
     val nParts = rdd.getNumPartitions
+
     val conf = VCFFileFormat.hadoopConfWithBGZ(data.sparkSession.sparkContext.hadoopConfiguration)
     val serializableConf = new SerializableConfiguration(conf)
     val firstNonemptyPartition =
       rdd.mapPartitions(iter => Iterator(iter.nonEmpty)).collect.indexOf(true)
 
-    if (firstNonemptyPartition == -1 && (!options.contains(VCFFileWriter.VCF_HEADER_KEY) ||
-      options
-        .get(VCFFileWriter.VCF_HEADER_KEY)
-        .contains(VCFFileWriter.INFER_HEADER))) {
-      throw new SparkException("Cannot infer header for empty VCF.")
+    if (firstNonemptyPartition == -1) {
+      if (!options.contains(VCFFileWriter.VCF_HEADER_KEY) ||
+        options
+          .get(VCFFileWriter.VCF_HEADER_KEY)
+          .contains(VCFFileWriter.INFER_HEADER)) {
+        throw new SparkException("Cannot infer header for empty VCF.")
+      }
+      if (nParts == 0) {
```

Comment:
I think we can throw this earlier (before we check for non-empty partitions).

Reply:
Moved up